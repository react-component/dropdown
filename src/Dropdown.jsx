import React, { Component } from 'react';
import PropTypes from 'prop-types';
import ReactDOM from 'react-dom';
import Trigger from 'rc-trigger';
import contains from 'rc-util/lib/Dom/contains';
import addEventListener from 'rc-util/lib/Dom/addEventListener';
import placements from './placements';

export default class Dropdown extends Component {
  static propTypes = {
    minOverlayWidthMatchTrigger: PropTypes.bool,
    onVisibleChange: PropTypes.func,
    onOverlayClick: PropTypes.func,
    prefixCls: PropTypes.string,
    children: PropTypes.any,
    transitionName: PropTypes.string,
    overlayClassName: PropTypes.string,
    animation: PropTypes.any,
    align: PropTypes.object,
    overlayStyle: PropTypes.object,
    placement: PropTypes.string,
    overlay: PropTypes.node,
    trigger: PropTypes.array,
    showAction: PropTypes.array,
    hideAction: PropTypes.array,
    getPopupContainer: PropTypes.func,
    visible: PropTypes.bool,
    defaultVisible: PropTypes.bool,
  };

  static defaultProps = {
    minOverlayWidthMatchTrigger: true,
    prefixCls: 'rc-dropdown',
    trigger: ['hover'],
    showAction: [],
    hideAction: [],
    overlayClassName: '',
    overlayStyle: {},
    defaultVisible: false,
    onVisibleChange() {
    },
    placement: 'bottomLeft',
  }

  constructor(props) {
    super(props);
    if ('visible' in props) {
      this.state = {
        visible: props.visible,
      };
    } else {
      this.state = {
        visible: props.defaultVisible,
      };
    }

    if (props.trigger.indexOf('contextMenu') !== -1) {
      addEventListener(document, 'mousedown', this.onDocumentClick);
      addEventListener(window, 'blur', this.onWindowBlur);
    }
  }

  componentWillReceiveProps({ visible }) {
    if (visible !== undefined) {
      this.setState({
        visible,
      });
    }
  }

  onClick = (e) => {
    const props = this.props;
    const overlayProps = props.overlay.props;
    // do no call onVisibleChange, if you need click to hide, use onClick and control visible
    if (!('visible' in props)) {
      this.setState({
        visible: false,
      });
    }
    if (props.onOverlayClick) {
      props.onOverlayClick(e);
    }
    if (overlayProps.onClick) {
      overlayProps.onClick(e);
    }
  }

  onContextMenu = (e) => {
    e.preventDefault();
    this.setState({
      visible: true,
    });
  }

  onDocumentClick = (e) => {
    // ContextMenu behavior
    if (!contains(this.trigger.getPopupDomNode(), e.target)) {
      this.setState({
        visible: false,
      });
    }
  }

  onWindowBlur = () => {
    // ContextMenu behavior
    this.setState({
      visible: false,
    });
  }

  onVisibleChange = (visible) => {
    const props = this.props;
    if (!('visible' in props)) {
      this.setState({
        visible,
      });
    }
    props.onVisibleChange(visible);
  }

  getMenuElement() {
    const { overlay, prefixCls } = this.props;
    const extraOverlayProps = {
      prefixCls: `${prefixCls}-menu`,
      onClick: this.onClick,
    };
    if (typeof overlay.type === 'string') {
      delete extraOverlayProps.prefixCls;
    }
    return React.cloneElement(overlay, extraOverlayProps);
  }

  getPopupDomNode() {
    return this.trigger.getPopupDomNode();
  }

  afterVisibleChange = (visible) => {
    if (visible && this.props.minOverlayWidthMatchTrigger) {
      const overlayNode = this.getPopupDomNode();
      const rootNode = ReactDOM.findDOMNode(this);
      if (rootNode && overlayNode && rootNode.offsetWidth > overlayNode.offsetWidth) {
        overlayNode.style.width = `${rootNode.offsetWidth}px`;
        if (this.trigger &&
            this.trigger._component &&
            this.trigger._component.alignInstance) {
          this.trigger._component.alignInstance.forceAlign();
        }
      }
    }
  }

  saveTrigger = (node) => {
    this.trigger = node;
  }

  render() {
    const {
      prefixCls, children,
      transitionName, animation,
      align, placement, getPopupContainer,
      showAction, hideAction,
      overlayClassName, overlayStyle,
      trigger, ...otherProps,
    } = this.props;
    return (
      <div onContextMenu={this.onContextMenu}>
        <Trigger
          {...otherProps}
          prefixCls={prefixCls}
          ref={this.saveTrigger}
          popupClassName={overlayClassName}
          popupStyle={overlayStyle}
          builtinPlacements={placements}
          action={trigger}
          showAction={showAction}
          hideAction={hideAction}
          popupPlacement={placement}
          popupAlign={align}
          popupTransitionName={transitionName}
          popupAnimation={animation}
          popupVisible={this.state.visible}
          afterPopupVisibleChange={this.afterVisibleChange}
          popup={this.getMenuElement()}
          onPopupVisibleChange={this.onVisibleChange}
          getPopupContainer={getPopupContainer}
        >
          {children}
        </Trigger>
      </div>
    );
  }
}
