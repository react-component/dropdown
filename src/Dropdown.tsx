import type { TriggerProps } from '@rc-component/trigger';
import Trigger from '@rc-component/trigger';
import type {
  ActionType,
  AlignType,
  AnimationType,
  BuildInPlacements,
} from '@rc-component/trigger/lib/interface';
import classNames from 'classnames';
import { composeRef, supportRef } from 'rc-util/lib/ref';
import type { ReactElement } from 'react';
import React from 'react';
import useAccessibility from './hooks/useAccessibility';
import Overlay from './Overlay';
import Placements from './placements';

export interface DropdownProps
  extends Pick<
    TriggerProps,
    | 'getPopupContainer'
    | 'children'
    | 'mouseEnterDelay'
    | 'mouseLeaveDelay'
    | 'onPopupAlign'
    | 'builtinPlacements'
    | 'autoDestroy'
  > {
  minOverlayWidthMatchTrigger?: boolean;
  arrow?: boolean;
  onVisibleChange?: (visible: boolean) => void;
  onOverlayClick?: (e: Event) => void;
  prefixCls?: string;
  transitionName?: string;
  overlayClassName?: string;
  openClassName?: string;
  animation?: AnimationType;
  align?: AlignType;
  overlayStyle?: React.CSSProperties;
  placement?: keyof typeof Placements;
  placements?: BuildInPlacements;
  overlay?: (() => React.ReactElement) | React.ReactElement;
  trigger?: ActionType | ActionType[];
  alignPoint?: boolean;
  showAction?: ActionType[];
  hideAction?: ActionType[];
  visible?: boolean;
  autoFocus?: boolean;
}

function Dropdown(props: DropdownProps, ref) {
  const {
    arrow = false,
    prefixCls = 'rc-dropdown',
    transitionName,
    animation,
    align,
    placement = 'bottomLeft',
    placements = Placements,
    getPopupContainer,
    showAction,
    hideAction,
    overlayClassName,
    overlayStyle,
    visible,
    trigger = ['hover'],
    autoFocus,
    overlay,
    children,
    onVisibleChange,
    ...otherProps
  } = props;

  const [triggerVisible, setTriggerVisible] = React.useState<boolean>();
  const mergedVisible = 'visible' in props ? visible : triggerVisible;

  const triggerRef = React.useRef(null);
  const overlayRef = React.useRef(null);
  const childRef = React.useRef(null);
  React.useImperativeHandle(ref, () => triggerRef.current);

  const handleVisibleChange = (newVisible: boolean) => {
    setTriggerVisible(newVisible);
    onVisibleChange?.(newVisible);
  };

  useAccessibility({
    visible: mergedVisible,
    triggerRef: childRef,
    onVisibleChange: handleVisibleChange,
    autoFocus,
    overlayRef,
  });

  const onClick = (e) => {
    const { onOverlayClick } = props;
    setTriggerVisible(false);

    if (onOverlayClick) {
      onOverlayClick(e);
    }
  };

  const getMenuElement = () => (
    <Overlay
      ref={overlayRef}
      overlay={overlay}
      prefixCls={prefixCls}
      arrow={arrow}
    />
  );

  const getMenuElementOrLambda = () => {
    if (typeof overlay === 'function') {
      return getMenuElement;
    }
    return getMenuElement();
  };

  const getMinOverlayWidthMatchTrigger = () => {
    const { minOverlayWidthMatchTrigger, alignPoint } = props;
    if ('minOverlayWidthMatchTrigger' in props) {
      return minOverlayWidthMatchTrigger;
    }

    return !alignPoint;
  };

  const getOpenClassName = () => {
    const { openClassName } = props;
    if (openClassName !== undefined) {
      return openClassName;
    }
    return `${prefixCls}-open`;
  };

  const childrenNode = React.cloneElement(children, {
    className: classNames(
      children.props?.className,
      mergedVisible && getOpenClassName(),
    ),
    ref: supportRef(children)
      ? composeRef(
          childRef,
          (children as ReactElement & { ref: React.Ref<HTMLElement> }).ref,
        )
      : undefined,
  });

  let triggerHideAction = hideAction;
  if (!triggerHideAction && trigger.indexOf('contextMenu') !== -1) {
    triggerHideAction = ['click'];
  }

  return (
    <Trigger
      builtinPlacements={placements}
      {...otherProps}
      prefixCls={prefixCls}
      ref={triggerRef}
      popupClassName={classNames(overlayClassName, {
        [`${prefixCls}-show-arrow`]: arrow,
      })}
      popupStyle={overlayStyle}
      action={trigger}
      showAction={showAction}
      hideAction={triggerHideAction}
      popupPlacement={placement}
      popupAlign={align}
      popupTransitionName={transitionName}
      popupAnimation={animation}
      popupVisible={mergedVisible}
      stretch={getMinOverlayWidthMatchTrigger() ? 'minWidth' : ''}
      popup={getMenuElementOrLambda()}
      onPopupVisibleChange={handleVisibleChange}
      onPopupClick={onClick}
      getPopupContainer={getPopupContainer}
    >
      {childrenNode}
    </Trigger>
  );
}

export default React.forwardRef(Dropdown);
