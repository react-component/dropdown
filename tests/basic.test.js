/* eslint-disable react/button-has-type,react/no-find-dom-node,react/no-render-return-value,object-shorthand,func-names,max-len */
import React from 'react';
import { mount } from 'enzyme';
import Menu, { Divider, Item as MenuItem } from 'rc-menu';
import { spyElementPrototypes } from 'rc-util/lib/test/domHook';
import { getPopupDomNode, sleep } from './utils';
import Dropdown from '../src';
import placements from '../src/placements';
import '../assets/index.less';

spyElementPrototypes(HTMLElement, {
  offsetParent: {
    get: () => document.body,
  },
  offsetLeft: {
    get: function () {
      return parseFloat(window.getComputedStyle(this).marginLeft) || 0;
    },
  },
  offsetTop: {
    get: function () {
      return parseFloat(window.getComputedStyle(this).marginTop) || 0;
    },
  },
  offsetHeight: {
    get: function () {
      return parseFloat(window.getComputedStyle(this).height) || 0;
    },
  },
  offsetWidth: {
    get: function () {
      return parseFloat(window.getComputedStyle(this).width) || 0;
    },
  },
});

describe('dropdown', () => {
  it('default visible', () => {
    const dropdown = mount(
      <Dropdown overlay={<div className="check-for-visible">Test</div>} visible>
        <button className="my-button">open</button>
      </Dropdown>,
    );
    expect(getPopupDomNode(dropdown) instanceof HTMLDivElement).toBeTruthy();
    expect(dropdown.find('.my-button').hasClass('rc-dropdown-open')).toBe(true);
  });

  it('supports constrolled visible prop', () => {
    const onVisibleChange = jest.fn();
    const dropdown = mount(
      <Dropdown
        overlay={<div className="check-for-visible">Test</div>}
        visible
        trigger={['click']}
        onVisibleChange={onVisibleChange}
      >
        <button className="my-button">open</button>
      </Dropdown>,
    );
    expect(getPopupDomNode(dropdown) instanceof HTMLDivElement).toBeTruthy();
    expect(dropdown.find('.my-button').hasClass('rc-dropdown-open')).toBe(true);

    dropdown.find('.my-button').simulate('click');
    expect(onVisibleChange).toHaveBeenCalledWith(false);
  });

  it('simply works', async () => {
    let clicked;
    let overlayClicked;

    function onClick({ key }) {
      clicked = key;
    }

    function onOverlayClick({ key }) {
      overlayClicked = key;
    }

    const menu = (
      <Menu style={{ width: 140 }} onClick={onClick}>
        <MenuItem key="1">
          <span className="my-menuitem">one</span>
        </MenuItem>
        <Divider />
        <MenuItem key="2">two</MenuItem>
      </Menu>
    );
    const dropdown = mount(
      <Dropdown trigger={['click']} overlay={menu} onOverlayClick={onOverlayClick}>
        <button className="my-button">open</button>
      </Dropdown>,
    );
    expect(dropdown.find('.my-button')).toBeTruthy();
    expect(dropdown.find('.rc-dropdown')).toBeTruthy();

    dropdown.find('.my-button').simulate('click');
    expect(clicked).toBeUndefined();
    expect(getPopupDomNode(dropdown).classList.contains('rc-dropdown-hidden')).toBe(false);

    dropdown.find('.my-menuitem').simulate('click');
    expect(clicked).toBe('1');
    expect(overlayClicked).toBe('1');
    expect(getPopupDomNode(dropdown).classList.contains('rc-dropdown-hidden')).toBe(true);
  });

  it('re-align works', async () => {
    const buttonStyle = { width: 600, height: 20, marginLeft: 100 };
    const menu = (
      <Menu>
        <MenuItem key="1">one</MenuItem>
      </Menu>
    );
    const dropdown = mount(
      <Dropdown trigger={['click']} placement="bottomRight" overlay={menu}>
        <button className="my-btn" style={buttonStyle}>
          open
        </button>
      </Dropdown>,
    );

    dropdown.find('.my-btn').simulate('click');
    await sleep(500);
    expect(getPopupDomNode(dropdown).getAttribute('style')).toEqual(
      expect.stringContaining(
        `left: -${999 - buttonStyle.width - placements.bottomLeft.offset[0]}px; top: -${
          999 - buttonStyle.height - placements.bottomLeft.offset[1]
        }px;`,
      ),
    );
  });

  // https://github.com/ant-design/ant-design/issues/9559
  it('should have correct menu width when switch from shorter menu to longer', async () => {
    class Example extends React.Component {
      state = { longList: true };

      getPopupDomNode() {
        return this.trigger.getPopupDomNode();
      }

      short = () => {
        this.setState({ longList: false });
      };

      long = () => {
        this.setState({ longList: true });
      };

      render() {
        const menuItems = [
          <MenuItem key="1">1st item</MenuItem>,
          <MenuItem key="2">2nd item</MenuItem>,
        ];
        if (this.state.longList) {
          menuItems.push(<MenuItem key="3">3rd LONG SUPER LONG item</MenuItem>);
        }
        return (
          <Dropdown
            trigger={['click']}
            ref={(node) => {
              this.trigger = node;
            }}
            overlay={<Menu>{menuItems}</Menu>}
          >
            <button>Actions 111</button>
          </Dropdown>
        );
      }
    }
    const dropdown = mount(<Example />);
    dropdown.find('button').simulate('click');
    await sleep(500);
    expect(getPopupDomNode(dropdown).getAttribute('style')).toEqual(
      expect.stringContaining(
        `left: -${999 - placements.bottomLeft.offset[0]}px; top: -${
          999 - placements.bottomLeft.offset[1]
        }px;`,
      ),
    );

    // Todo - offsetwidth
  });

  it('Test default minOverlayWidthMatchTrigger', async () => {
    const overlayWidth = 50;
    const overlay = <div style={{ width: overlayWidth }}>Test</div>;

    const dropdown = mount(
      <Dropdown trigger={['click']} overlay={overlay}>
        <button style={{ width: 100 }} className="my-button">
          open
        </button>
      </Dropdown>,
    );

    dropdown.find('.my-button').simulate('click');
    await sleep(500);
    expect(getPopupDomNode(dropdown).getAttribute('style')).toEqual(
      expect.stringContaining('min-width: 100px'),
    );
  });

  it('user pass minOverlayWidthMatchTrigger', async () => {
    const overlayWidth = 50;
    const overlay = <div style={{ width: overlayWidth }}>Test</div>;

    const dropdown = mount(
      <Dropdown trigger={['click']} overlay={overlay} minOverlayWidthMatchTrigger={false}>
        <button style={{ width: 100 }} className="my-button">
          open
        </button>
      </Dropdown>,
    );

    dropdown.find('.my-button').simulate('click');
    await sleep(500);
    expect(getPopupDomNode(dropdown).getAttribute('style')).not.toEqual(
      expect.stringContaining('min-width: 100px'),
    );
  });

  it('should support default openClassName', () => {
    const overlay = <div style={{ width: 50 }}>Test</div>;
    const dropdown = mount(
      <Dropdown trigger={['click']} overlay={overlay} minOverlayWidthMatchTrigger={false}>
        <button style={{ width: 100 }} className="my-button">
          open
        </button>
      </Dropdown>,
    );
    dropdown.find('.my-button').simulate('click');
    expect(dropdown.find('.my-button').prop('className')).toBe('my-button rc-dropdown-open');
    dropdown.find('.my-button').simulate('click');
    expect(dropdown.find('.my-button').prop('className')).toBe('my-button');
  });

  it('should support custom openClassName', async () => {
    const overlay = <div style={{ width: 50 }}>Test</div>;
    const dropdown = mount(
      <Dropdown
        trigger={['click']}
        overlay={overlay}
        minOverlayWidthMatchTrigger={false}
        openClassName="opened"
      >
        <button style={{ width: 100 }} className="my-button">
          open
        </button>
      </Dropdown>,
    );

    dropdown.find('.my-button').simulate('click');
    expect(dropdown.find('.my-button').prop('className')).toBe('my-button opened');
    dropdown.find('.my-button').simulate('click');
    expect(dropdown.find('.my-button').prop('className')).toBe('my-button');
  });

  it('overlay callback', async () => {
    const overlay = <div style={{ width: 50 }}>Test</div>;
    const dropdown = mount(
      <Dropdown trigger={['click']} overlay={() => overlay}>
        <button className="my-button">open</button>
      </Dropdown>,
    );

    dropdown.find('.my-button').simulate('click');
    expect(getPopupDomNode(dropdown).classList.contains('rc-dropdown-hidden')).toBe(false);
  });

  it('should support arrow', async () => {
    const overlay = <div style={{ width: 50 }}>Test</div>;
    const dropdown = mount(
      <Dropdown arrow overlay={overlay} trigger={['click']}>
        <button style={{ width: 100 }} className="my-button">
          open
        </button>
      </Dropdown>,
    );

    dropdown.find('.my-button').simulate('click');
    await sleep(500);
    expect(getPopupDomNode(dropdown).classList.contains('rc-dropdown-show-arrow')).toBe(true);
    expect(
      getPopupDomNode(dropdown).firstElementChild.classList.contains('rc-dropdown-arrow'),
    ).toBe(true);
  });
});
