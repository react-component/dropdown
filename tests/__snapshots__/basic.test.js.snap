// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dropdown simply works 1`] = `
Array [
  <button
    class="my-button rc-dropdown-open"
  >
    open
  </button>,
  <div>
    <div
      class="rc-dropdown"
      style="opacity: 0;"
    >
      <ul
        class="rc-dropdown-menu rc-dropdown-menu-root rc-dropdown-menu-vertical"
        data-dropdown-inject="true"
        data-menu-list="true"
        role="menu"
        style="width: 140px;"
        tabindex="0"
      >
        <li
          class="rc-dropdown-menu-item"
          data-menu-id="rc-menu-uuid-test-1"
          role="menuitem"
          tabindex="-1"
        >
          <span
            class="my-menuitem"
          >
            one
          </span>
        </li>
        <li
          class="rc-dropdown-menu-item-divider"
        />
        <li
          class="rc-dropdown-menu-item"
          data-menu-id="rc-menu-uuid-test-2"
          role="menuitem"
          tabindex="-1"
        >
          two
        </li>
      </ul>
      <div
        aria-hidden="true"
        style="display: none;"
      />
    </div>
  </div>,
]
`;
